memberlist_test.go

```diff
@@ -1316,197 +1364,350 @@ func TestMemberlist_conflictDelegate(t *testing.T) {
-	addr2 := getBindAddr()
-	conf2.Name = addr2.String()
-	conf2.BindAddr = addr2.String()
-	conf2.BindPort = m1.config.BindPort
-	conf2.GossipInterval = time.Millisecond
-	conf2.SecretKey = []byte("Hi16ZXu2lNCRVwtr20khAg==")
-	conf2.GossipVerifyIncoming = false
-	conf2.GossipVerifyOutgoing = false
-	conf2.Logger = testLogger(t)
-
-	m2, err := Create(conf2)
+		if m.NumMembers() != expected {
+			failf("%s expected to have %d members but had: %v", m.config.Name, expected, m.Members())
+		}
+	})
+}
+
+func isPortFree(t *testing.T, addr string, port int) error {
+	t.Helper()
+
+	ip := net.ParseIP(addr)
+	tcpAddr := &net.TCPAddr{IP: ip, Port: port}
+	tcpLn, err := net.ListenTCP("tcp", tcpAddr)
 	if err != nil {
-		t.Fatalf("unexpected err: %s", err)
+		return err
 	}
-	defer m2.Shutdown()
-
-	// Join the second node. m1 has no encryption while m2 has encryption configured and
-	// can receive encrypted gossip, but will not encrypt outgoing gossip.
-	num, err := m2.Join([]string{m1.config.BindAddr})
-	if num != 1 {
-		t.Fatalf("unexpected 1: %d", num)
+	if err := tcpLn.Close(); err != nil {
+		return err
 	}
+
+	udpAddr := &net.UDPAddr{IP: ip, Port: port}
+	udpLn, err := net.ListenUDP("udp", udpAddr)
 	if err != nil {
-		t.Fatalf("unexpected err: %s", err)
+		return err
 	}
 
-	// Check the hosts
-	if len(m2.Members()) != 2 {
-		t.Fatalf("should have 2 nodes! %v", m2.Members())
-	}
-	if m2.estNumNodes() != 2 {
-		t.Fatalf("should have 2 nodes! %v", m2.Members())
-	}
+	return udpLn.Close()
+}
 
-	// Leave with the first node
-	m1.Leave(time.Second)
+func waitUntilPortIsFree(t *testing.T, m *Memberlist) {
+	t.Helper()
 
-	// Wait for leave
-	time.Sleep(10 * time.Millisecond)
+	// wait until we know for certain that m1 is dead dead
+	addr := m.config.BindAddr
+	port := m.config.BindPort
 
-	// Create a third node that has the second stage of gossip transition settings
-	conf3 := DefaultLANConfig()
-	addr3 := getBindAddr()
-	conf3.Name = addr3.String()
-	conf3.BindAddr = addr3.String()
-	conf3.BindPort = m1.config.BindPort
-	conf3.GossipInterval = time.Millisecond
-	conf3.SecretKey = conf2.SecretKey
-	conf3.GossipVerifyIncoming = false
-	conf3.Logger = testLogger(t)
-
-	m3, err := Create(conf3)
-	if err != nil {
-		t.Fatalf("unexpected err: %s", err)
-	}
-	defer m3.Shutdown()
+	retry(t, 15, 250*time.Millisecond, func(failf func(string, ...interface{})) {
+		t.Helper()
 
-	// Join the third node to the second node. At this step, both nodes have encryption
-	// configured but only m3 is sending encrypted gossip.
-	num, err = m3.Join([]string{m2.config.BindAddr})
-	if num != 1 {
-		t.Fatalf("unexpected 1: %d", num)
-	}
-	if err != nil {
-		t.Fatalf("unexpected err: %s", err)
-	}
+		if err := isPortFree(t, addr, port); err != nil {
+			failf("%s port is not yet free", m.config.Name)
+		}
+	})
+}
 
-	// Check the hosts
-	if len(m3.Members()) != 2 {
-		t.Fatalf("should have 2 nodes! %v", m3.Members())
+// This test should follow the recommended upgrade guide:
```

Comment:
The prior version of this test technically _worked_ but was testing an undesirable situation that isn't part of the gossip encryption upgrade guide.  There were 4 nodes (same as the new test):
* A: no encryption
* B: encryption, no-verify
* C: encryption, verify-out
* D: encryption, verify-in, verify-out

There would only ever be 2 full members of the pool at each iteration of the test, but the Left members A and B would remain gossiping.  As A was incapable of understanding encrypted messages, both C and D could not directly communicate with that node, and A had to rely upon gossip forwarded from B, which was still up.

The new version of the test instead tries to follow the spirit of the Official Procedure and manipulates a two-node cluster, transitioning it to the various approved states so that the behavior is more regular and doesn't cause any full or partial network partitions that make the test even more timing dependent.